chainstate-storage/src/internal/mod.rs

```diff
@@ -374,19 +374,19 @@ impl<'a, Tx: traits::GetMapMut<'a, Schema>> StoreTx<Tx> {
 impl<T: traits::TransactionRw<Error = storage::Error>> traits::TransactionRw for StoreTx<T> {
     type Error = crate::Error;
 
-    fn commit(self) -> crate::Result<()> {
+    fn commit(&mut self) -> crate::Result<()> {
```

Comment:
Yep. In agreement here

Reply:
Any better ideas how to make it object safe or achieve the required abstraction in some other way?

Reply:
Can you please elaborate more on the problem? From what I'm seeing, there are no big changes in this PR that necessitate this change. I think we should fight a little more before giving up on this awesome feature. I believe with the right set of wrappers, everything is possible.

Reply:
> Can you please elaborate more on the problem?

The problem is that when you use a trait object by wrappig it in `dyn`, for example when you do `dyn TransactionRo`, information about the size of the object is lost. Without size, compiler no longer knows how to move the obejct. So using `dyn Trait` is only allowed if the said `Trait` does not move `self`.

>  From what I'm seeing, there are no big changes in this PR that necessitate this change.

Not in this PR. I have some code locally that uses trait object to abstract away chainstate from having to use generics using the likes of `Box<dyn BlockchainStorage>`, `Box<dyn TransactionRo>`, etc. What I like about that approach is that it is quite transparent to the rest of the system. I managed to make chainstate use the dynamic (backend agnostic) Store type instead of inmemory, without single extra line of code having to be modified in chainstate.

> I think we should fight a little more before giving up on this awesome feature. I believe with the right set of wrappers, everything is possible.

Therese are the options I can see to make chainstate backend-agnostic:

1. Use `dyn` trait objects for Storage. By using the `Transaction{Ro,Rw}::run` methods (with lambdas), the fact that commit/abort don't consume the object is mitigated somewhat. It also ensures that one of them is called eventually.
   DIsadvantages: Requires changes from this PR (making transactions object-safe).
   Advantages: completely transparent to chainstate as described above.
2. Use an enum with a variant for each backend type. This was my fallback option if using `dyn` fails. I don't like it.
   Advantages: traits don't need to be object safe; still completely transparent to chainstate.
   Disadvantages: Every call has to be dispatched using `match`, making for some ugly boilerplate. All supported backends have to be known ahead of time, reducing flexibility and introducing more coupling between the components. Adding a new backend (or mock interface) will be a bit of a pain.
3. Make chainstate generic over storage backend, do the dynamic dispatch at the `ChanstateInterface` level (which already is object safe). Maybe the preferred option in the log run.
    DIsadvantages: Almost everything in chainstate would have to be generic, implemented against `BlockchainStorage{,Read,Write}` traits. That can be a quite disruptive change which would conflict with many things. I believe some parts already are but it should be done eventually to support mocking storage out too. Also the code in cahinstate would have to be compiled multiple times, once for each storage backend.
   Risks: I am not 100% sure this will avoid the object safety changes in transaction traits. My intuition is it should.
   Advantages: Lowest overhead in terms of runtime performance. Internal chainstate interfaces will be better specified by using traits rather than concrete types.

Reply:
There's absolutely no question that generics is the way to go. However, having said that.

I'd like to point out that a wrapper around a dyn should make it movable. Knowing the size of an object isn't a necessity to move it. Why? Because moving, at low level, is just copying the pointer and discarding another in another stack frame. It should not pose a problem. See the UtxosCache for example. I don't think it's not movable because it has a dyn reference. Maybe that's my ignorance in rust or there's something I ain't seeing. But either way, if this is really impossible, then we go the generics route.

Reply:
> I'd like to point out that a wrapper around a dyn should make it movable. Knowing the size of an object isn't a necessity to move it. Why? Because moving, at low level, is just copying the pointer and discarding another in another stack frame. It should not pose a problem. See the UtxosCache for example. I don't think it's not movable because it has a dyn reference. Maybe that's my ignorance in rust or there's something I ain't seeing. But either way, if this is really impossible, then we go the generics route.

Well, you can't have `dyn Trait` where methods consume `Self` no matter how many wrappers you slap on self. You can create a wrapper over `Box<dyn Trait>` that provides the "consuming" interface. However, the trait would still have to use `&mut self` in order to be allowed to create trait objects with it. Here is the best I was able to come up with.

You have two interfaces. One provides the consuming API (simplifying a bit):

```rust
trait DbTx: Sized {
    fn commit(self);
}
```

One is object safe:
```rust
trait DynDbTx {
    fn dyn_commit(&mut self);
}
```

The `DbTx` trait is the public interface used by the clients while `DynDbTx` is what gets implemented by storage backends. A wrapper type can be introduced to bridge from `DynDbTx` to `DbTx` for things that also happen to be `Sized`:

```rust
struct UnDyn<T>(T);

impl<T: DynDbTx + Sized> DbTx for UnDyn<T> {
    fn commit(self) {
        self.0.dyn_commit()
    }
}
```
Then, together with the following boilerplate impl:
```rust
impl<T: DynDbTx> impl DynDbTx for Box<T> { /* trivial */ }
```
the `UnDyn<Box<dyn DynDbTx>>` implements `DbTx`. That allows for turning the `dyn DynDbTx` trait object into something implementing the consuming interface. Another wrapper could be slapped on top to hide all the internal wrapping and boxing layers. It's all a bit messy, especially the need to have two almost identical interfaces and the number of layers required.

Anyway I am currently exploring the generics alternative so this is not of too much relevance right now. It might be useful to remember this pattern in case the issue of providing a consuming interface to trait objects comes up again.

Hope this illuminates the issue a bit.

Reply:
Unfortunately, I can't roll up my sleeves and join in full capacity. Let's see where generics will take us for now. I'll look into this further when I can.

Reply:
I think the crux of the issue is the following:
* An object safe trait is **not allowed** to have `Self: Sized`
* A method with the `Self` type (taken by value) implicitly imposes `Self: Sized`

Can't have both simultaneously in one trait. I am not sure it's possible to do much better than the dual interface approach mentioned above. 

Relevant reference section is at https://doc.rust-lang.org/reference/items/traits.html#object-safety